ckb-combine-lock-common/src/ckb_auth.rs

```diff
@@ -0,0 +1,239 @@
+}
+
+impl Into<u8> for AuthAlgorithmIdType {
+    fn into(self) -> u8 {
+        self as u8
+    }
+}
+
+impl TryFrom<u8> for AuthAlgorithmIdType {
+    type Error = CkbAuthError;
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        if (value >= AuthAlgorithmIdType::Ckb.into()
+            && value <= AuthAlgorithmIdType::Iso97962.into())
+            || value == AuthAlgorithmIdType::OwnerLock.into()
+        {
+            Ok(unsafe { transmute(value) })
+        } else {
+            Err(CkbAuthError::UnknowAlgorithmID)
+        }
+    }
+}
+
+pub struct CkbAuthType {
+    pub algorithm_id: AuthAlgorithmIdType,
+    pub pubkey_hash: [u8; 20],
+}
+
+pub enum EntryCategoryType {
+    Exec = 0,
+    DynamicLinking = 1,
+}
+
+impl TryFrom<u8> for EntryCategoryType {
+    type Error = CkbAuthError;
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Self::Exec),
+            1 => Ok(Self::DynamicLinking),
+            _ => Err(CkbAuthError::EncodeArgs),
+        }
+    }
+}
+
+pub struct CkbEntryType {
+    pub code_hash: [u8; 32],
+    pub hash_type: ScriptHashType,
+    pub entry_category: EntryCategoryType,
+}
+
+pub fn ckb_auth(
+    entry: &CkbEntryType,
+    id: &CkbAuthType,
+    signature: &[u8],
+    message: &[u8; 32],
+) -> Result<(), CkbAuthError> {
+    match entry.entry_category {
+        EntryCategoryType::Exec => ckb_auth_exec(entry, id, signature, message),
+        EntryCategoryType::DynamicLinking => ckb_auth_dl(entry, id, signature, message),
+    }
+}
+
+fn ckb_auth_exec(
+    entry: &CkbEntryType,
+    id: &CkbAuthType,
+    signature: &[u8],
+    message: &[u8; 32],
+) -> Result<(), CkbAuthError> {
+    let args = CString::new(format!(
+        "{}:{:02X?}:{:02X?}:{}:{}:{}",
+        encode(&entry.code_hash),
+        entry.hash_type as u8,
+        id.algorithm_id.clone() as u8,
+        encode(signature),
+        encode(message),
+        encode(id.pubkey_hash)
+    ))?;
+
+    // log!("args: {:?}", args);
+    exec_cell(&entry.code_hash, entry.hash_type, 0, 0, &[args.as_c_str()])?;
+    Ok(())
+}
+
+type CkbAuthValidate = unsafe extern "C" fn(
+    auth_algorithm_id: u8,
+    signature: *const u8,
+    signature_size: u32,
+    message: *const u8,
+    message_size: u32,
+    pubkey_hash: *mut u8,
+    pubkey_hash_size: u32,
+) -> i32;
+
+struct DynamicLinkingContext {
+    _context: CKBDLContext<[u8; 256 * 1024]>,
+    ckb_auth_validate: Symbol<CkbAuthValidate>,
+}
+static mut G_DL_CONTEXT: Option<DynamicLinkingContext> = None;
+
+impl DynamicLinkingContext {
+    fn get() -> Result<&'static Self, CkbAuthError> {
+        unsafe {
+            if G_DL_CONTEXT.is_some() {
+                Ok(G_DL_CONTEXT.as_ref().unwrap())
+            } else {
+                Err(CkbAuthError::DynamicLinkingUninit)
+            }
+        }
+    }
+
+    fn new(code_hash: &[u8; 32], hash_type: ScriptHashType) -> Result<&'static Self, CkbAuthError> {
```

Comment:
Consider the following scenario: multiple dynamic libraries, whether they are the same or distinct, are being loaded. Here is the reference implementation: https://github.com/nervosnetwork/ckb-production-scripts/blob/abdcb117b512e35910fa8e30241a7a354e5cacf0/c/xudt_rce.c#L91-L119.